modules/core/include/opencv2/core/fast_math.hpp

```diff
@@ -133,7 +133,8 @@
         defined(__x86_64__) || defined(__i686__) \
         || defined(__arm__) \
         || defined(__PPC64__) \
-    )
+    ) \
+    && !defined(__FAST_MATH__)
```

Comment:
> OPENCV_USE_FASTMATH_BUILTINS

It doesn't make sense to disable all builtins for `fast-math` (just slow down anything).

Related problem is related to `cvIsNan()` only (probably no problem at all with `cvIsNaN()` as there is still no test).

Also this check doesn't handle other cases like:

```
  #elif defined __PPC64__ && defined _ARCH_PWR9 && defined(scalar_test_data_class)
    #define CV_INLINE_ISNAN_DBL(value) return scalar_test_data_class(value, 0x40);
    #define CV_INLINE_ISNAN_FLT(value) CV_INLINE_ISNAN_DBL(value)
  #endif
```

At first we need to start from tests.

```.diff
 CV_INLINE int cvIsNaN( double value )
 {
-#if defined CV_INLINE_ISNAN_DBL
+#if defined(CV_INLINE_ISNAN_DBL) && !defined(__FAST_MATH__)
     CV_INLINE_ISNAN_DBL(value);
 #else
     Cv64suf ieee754;

 CV_INLINE int cvIsNaN( float value )
 {
-#if defined CV_INLINE_ISNAN_FLT
+#if defined(CV_INLINE_ISNAN_FLT) && !defined(__FAST_MATH__)
     CV_INLINE_ISNAN_FLT(value);
 #else
     Cv32suf ieee754;

```

Some simple test (to avoid reversing of modified implementations):

```
TEST(Core_FastMath, cvIsNaN)
{
    float nf = std::numeric_limits<float>::quiet_NaN();
    EXPECT_TRUE(cvIsNaN(nf));
    float nd = std::numeric_limits<double>::quiet_NaN();
    EXPECT_TRUE(cvIsNaN(nd));
}
```

---

BTW, existed `isinf()` test fails too. Perhaps we need to add similar checks here too.

Reply:
Anyway own integer-based check is really slow, we should try to use CPU instructions for that.

Looks like compiler optimize NAN checks to `false` constant in fast-math mode.

Some investigations including `-fno-finite-math-only`:
- https://godbolt.org/z/9T4eGf6sd